durabletask/task.py

```diff
@@ -313,6 +296,92 @@ def get_completed_tasks(self) -> int:
         return self._completed_tasks
 
 
+class CompletableTask(Task[T]):
+
+    def __init__(self):
+        super().__init__()
+        self._retryable_parent = None
+
+    def complete(self, result: T):
+        if self._is_complete:
+            raise ValueError('The task has already completed.')
+        self._result = result
+        self._is_complete = True
+        if self._parent is not None:
+            self._parent.on_child_completed(self)
+
+    def fail(self, message: str, details: pb.TaskFailureDetails):
+        if self._is_complete:
+            raise ValueError('The task has already completed.')
+        self._exception = TaskFailedError(message, details)
+        self._is_complete = True
+        if self._parent is not None:
+            self._parent.on_child_completed(self)
+
+
+class RetryableTask(CompletableTask[T]):
+
+    def __init__(self, retry_policy: RetryPolicy, action: pb.OrchestratorAction,
+                 is_sub_orch: bool) -> None:
+        super().__init__()
+        self._action = action
+        self._retry_policy = retry_policy
+        self._retry_count = 0
+        self._first_attempt: datetime
+        self._is_sub_orch = is_sub_orch
+
+    def try_completion(self) -> bool:
+        if self._retry_count >= self._retry_policy.max_number_of_attempts - 1:
+            return True
+        else:
+            return False
+    
+    def set_first_attempt(self, first_attempt: datetime) -> None:
+        self._first_attempt = first_attempt
+
+    def increment_retry_count(self) -> None:
+        self._retry_count += 1
+    
+    def compute_next_delay(self) -> timedelta:
+        next_delay = timedelta.min
+        retry_expiration: datetime = datetime.max
+        if self._retry_policy.retry_timeout is not None and self._retry_policy.retry_timeout != datetime.max:
+            retry_expiration = self._first_attempt + self._retry_policy.retry_timeout
+        
+        if self._retry_policy.backoff_coefficient is None:
+            backoff_coefficient = 1.0
+        else:
+            backoff_coefficient = self._retry_policy.backoff_coefficient
+
+        if datetime.utcnow() < retry_expiration:
+            next_delay_f = math.pow(backoff_coefficient, self._retry_count) * self._retry_policy.first_retry_interval.total_seconds()
+
+            if self._retry_policy.max_retry_interval is not None:
+                next_delay_f = min(next_delay_f, self._retry_policy.max_retry_interval.total_seconds())
+                return timedelta(seconds=next_delay_f)
+
+        return next_delay
```

Comment:
It looks like you're returning `timedelta.min` if the retry expiration has passed. I wonder if it would be more intuitive to return `None` in that case as a more clear signal that we cannot delay any further.

Reply:
returned None and checked accordingly.